lib/web_of_science/contributions.rb

```diff
@@ -34,40 +34,43 @@ def author_contribution(author, uid)
     # @param [Publication]
     # @return [Contribution]
     def find_or_create_contribution(author, publication)
-      contrib = publication.contributions.find_or_initialize_by(author_id: author.id)
-      create_contribution(author, contrib) unless contrib.persisted?
-      contrib
-    end
-
-    # Save a new contribution to a WOS-UID publication for author.
-    # @param author [Author]
-    # @param contrib [Contribution]
-    # @return [Boolean]
-    def create_contribution(author, contrib)
-      contrib.assign_attributes(
-        cap_profile_id: author.cap_profile_id,
-        featured: false, status: 'new', visibility: 'private'
-      )
-      contrib.save!
-      # Add the pub_hash[:authorship] data to the publication
-      contrib.publication.pubhash_needs_update!
-      contrib.publication.save!
+      publication.contributions.find_or_create_by!(author_id: author.id) do |contrib|
+        contrib.assign_attributes(
+          cap_profile_id: author.cap_profile_id,
+          featured: false, status: 'new', visibility: 'private'
+        )
+        publication.pubhash_needs_update! # Add to pub_hash[:authorship]
+        publication.save! # contrib.save! not needed
+      end
     rescue ActiveRecord::ActiveRecordError => err
-      message = "Failed to create contribution for author: #{author.id}, pub: #{contrib.publication.id}"
+      message = "Failed to find/create contribution for author: #{author.id}, pub: #{publication.id}"
       NotificationManager.error(err, message, self)
-      false
     end
 
     # Does record have a contribution for this author? (based on matching PublicationIdentifiers)
     # Note: must use unique identifiers, don't use ISSN or similar series level identifiers
+    # We search for all PubIDs at once instead of serial queries.  No need to hit the same table multiple times.
     # @param [Author] author
     # @param [WebOfScience::Record] record
-    # @return [Boolean]
-    def found_contribution?(author, record)
-      contribution_by_identifier?(author, 'WosUID', record.uid) ||
-        contribution_by_identifier?(author, 'WosItemID', record.wos_item_id) ||
-        contribution_by_identifier?(author, 'doi', record.doi) ||
-        contribution_by_identifier?(author, 'pmid', record.pmid)
+    # @return [::Contribution, nil] a matched or newly minted Contribution
+    def matching_contribution(author, record)
+      pub = Publication.joins(:publication_identifiers).where(
+        "publication_identifiers.identifier_value IS NOT NULL AND (
```

Comment:
This clause would not be necessary if we had the desired validations/constraints on `publication_identifiers`.